File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum;
    public int solve(TreeNode root){
        if(root==null){
            return 0;
        }
        int left = solve(root.left);
        left = (left<0)?0:left;
        int right = solve(root.right);
        right = (right<0)?0:right;
        
        maxSum = Math.max(maxSum, left+right+root.val);
        
        return Math.max(left,right)+root.val;
    }
    public int maxPathSum(TreeNode root) {
        maxSum=Integer.MIN_VALUE;
        solve(root);
        return maxSum;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* root, int& maxSum){
        if(root==NULL){
            return 0;
        }
        int left = solve(root->left, maxSum);
        left = (left<0)?0:left;
        int right = solve(root->right, maxSum);
        right = (right<0)?0:right;
        
        maxSum = max(maxSum, left+right+root->val);
        
        return max(left, right)+root->val;
    }
    int maxPathSum(TreeNode* root) {
        int maxSum=INT_MIN;
        solve(root, maxSum);
        return maxSum;
    }
};